iocore/net/quic/QUICStream.cc

```diff
@@ -409,3 +417,63 @@ QUICStream::largest_offset_sent()
 {
   return this->_remote_flow_controller->current_offset();
 }
+
+bool
+QUICStream::is_valid_fin_offset(QUICOffset last_offset) const
+{
+  // {sector 11.3. Stream Final Offset}
+  // Once a final offset for a stream is known, it cannot change.
+  // If a RST_STREAM or STREAM frame causes the final offset to change for a stream,
+  // an endpoint SHOULD respond with a FINAL_OFFSET_ERROR error (see Section 12).
+  // A receiver SHOULD treat receipt of data at or beyond the final offset as a
+  if (this->_last_offset == UINT64_MAX) {
+    return true;
+  }
+
+  if (this->_last_offset != last_offset) {
+    return false;
+  }
+
+  return true;
+}
+
+QUICErrorCode
+QUICStream::get_error_code() const
+{
+  return this->_rst_error_code;
+}
+
+bool
+QUICStream::is_remote_closed() const
+{
+  return this->_state.get() == QUICStreamState::State::half_closed_remote || this->_state.get() == QUICStreamState::State::closed;
+}
+
+bool
+QUICStream::is_local_closed() const
+{
+  return this->_state.get() == QUICStreamState::State::half_closed_local || this->_state.get() == QUICStreamState::State::closed;
+}
+
+QUICError
+QUICStream::_check_and_set_fin_offset(QUICOffset offset, bool set_last_offset)
+{
+  if (set_last_offset) {
+    if (!this->is_valid_fin_offset(offset)) {
+      return QUICError(QUICErrorClass::QUIC_TRANSPORT, QUICErrorCode::FINAL_OFFSET_ERROR);
+    }
+
+    this->_last_offset = offset;
```

Comment:
Do you mean final offset by `_last_offset`? Why don't you call it `_final_offset`?

Reply:
That's right, I'll fix it.